gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -572,10 +573,14 @@ local function SetSubRoles(plys, availableRoles, selectableRoles, selectedForced
 	local plysAmount = #plys
 	local availableRolesAmount = #availableRoles
 	local tmpSelectableRoles = table.Copy(selectableRoles)
+	-- plys are already shuffled, continue with existing priority
+	local newPlys = table.Copy(plys)
 
-	while plysAmount > 0 and availableRolesAmount > 0 do
-		local pick = math.random(plysAmount)
-		local ply = plys[pick]
+	for i = #newPlys, 1, -1 do
```

Comment:
The for loop here also has the wrong behaviour. It only goes through every player once. But it could be that some players are not assigned.